service/raft/raft_group_registry.cc

```diff
@@ -473,8 +476,24 @@ raft_server_with_timeouts::run_with_timeout(Op&& op, const char* op_name,
         if (!expiry.abort_source().abort_requested() || (as && as->abort_requested())) {
             throw;
         }
-        // TODO: improve error message (check for quorum loss etc)
-        const auto message = ::format("raft operation [{}] timed out", op_name);
+        auto message = ::format("raft operation [{}] timed out", op_name);
+        {
+            auto fd = _registry.failure_detector();
+            unsigned voters_count = 0;
+            unsigned alive_voters_count = 0;
+            for (const auto& c: _server.get_configuration().current) {
+                if (c.can_vote) {
+                    ++voters_count;
+                    if (fd->is_alive(c.addr.id)) {
+                        ++alive_voters_count;
+                    }
+                }
+            }
+            if (alive_voters_count <= voters_count / 2) {
+                message += ::format(", no group0 quorum, total voters {}, alive voters {}",
```

Comment:
Nice.

Reply:
Although it would be better (performance-wise) if we didn't do a string addition here, but did the voters calculation before creating the message. Could matter if there's a batch of timeouts happening at the same time (maybe not with group 0, but in the future with multiple groups)

Reply:
> It's just one alloc, I don't think it matters.


Famous last words